gateway/ln-gateway/src/bin/cln_extension.rs

```diff
@@ -0,0 +1,330 @@
+    pub forward_amount: Amount,
+    pub outgoing_cltv_value: u32,
+    pub shared_secret: bitcoin_hashes::sha256::Hash,
+    pub next_onion: String,
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub struct HtlcAccepted {
+    pub htlc: Htlc,
+    pub onion: Onion,
+}
+
+pub struct ClnRpcClient {}
+
+#[allow(dead_code)]
+pub struct ClnRpcService {
+    client: Arc<Mutex<ClnRpc>>,
+    interceptor: Arc<Mutex<ClnHtlcInterceptor>>,
+    task_group: TaskGroup,
+    pub dir: PathBuf,
+}
+
+impl ClnRpcService {
+    pub async fn new() -> Result<Self, ClnRpcError> {
+        let interceptor = Arc::new(Mutex::new(ClnHtlcInterceptor::new()));
+
+        if let Some(plugin) = Builder::new(stdin(), stdout())
+            .option(options::ConfigOption::new(
+                "gateway-cfg",
+                // FIXME: cln_plugin doesn't support parameters without defaults
+                options::Value::String("default-dont-use".into()),
+                "gateway config directory",
+            ))
+            .hook(
+                "htlc_accepted",
+                |plugin: Plugin<Arc<Mutex<ClnHtlcInterceptor>>>, value: serde_json::Value| async move {
+                    // This callback needs to be `Sync`, so we use tokio::spawn
+                    let handle = tokio::spawn(async move {
+                        // Handle core-lightning "htlc_accepted" events
+                        // by passing the HTLC to the interceptor in the plugin state
+                        let payload: HtlcAccepted = serde_json::from_value(value)?;
+                        Ok(plugin.state().lock().await.intercept_htlc(payload).await)
+                    });
+                    handle.await?
+                },
+            )
+            .dynamic() // Allow reloading the plugin
+            .start(interceptor.clone())
+            .await?
+        {
+            let config = plugin.configuration();
+            let socket = PathBuf::from(config.lightning_dir).join(config.rpc_file);
+            let client = ClnRpc::new(socket).await.expect("connect to ln_socket");
+
+            let dir = match plugin.option("gateway-cfg") {
+                Some(options::Value::String(workdir)) => {
+                    // FIXME: cln_plugin doesn't yet support optional parameters
+                    if &workdir == "default-dont-use" {
+                        panic!("gateway-cfg option missing")
+                    } else {
+                        PathBuf::from(workdir)
+                    }
+                }
+                _ => unreachable!(),
+            };
+
+            Ok(Self {
+                client: Arc::new(Mutex::new(client)),
+                task_group: TaskGroup::new(),
+                interceptor,
+                dir,
+            })
+        } else {
+            // TODO: Accurately define LightningError when building the plugin fails
+            Err(ClnRpcError::RpcServerError(LightningError(Some(0))))
```

Comment:
What does the number `0` represent here?

Reply:
Randomly picked a `LightningError` to report. This is inconsequential since the error is reported to the service builder.
Marked a todo above to accurately define these errors by context,. We can follow up with telemetry